doc/codegen/modular_codegen_design.md

```diff
@@ -0,0 +1,571 @@
+- **Current Location**: Embedded in CUDA code generation
+
+#### 10. **TaskletLanguageLoweringPass**
+- **Purpose**: Convert Python/generic tasklets to target language (C++/CUDA/etc.)
+- **Input**: SDFG with tasklets
+- **Output**: SDFG with lowered tasklets
+- **Current Location**: Distributed across target generators
+
+### Phase 3: Code Generation Passes
+
+#### 10. **StateStructCreationPass**
+- **Purpose**: Generate state struct definitions for persistent data
+- **Input**: SDFG with allocation info
+- **Output**: pipeline_results["state_struct"] = {struct_def, struct_init}
+- **Current Location**: `DaCeCodeGenerator.generate_code()`
+
+#### 11. **MemletLoweringPass**
+- **Purpose**: Lower high-level memlets to explicit copy operations
+- **Input**: SDFG with target analysis
+- **Output**: SDFG with explicit copies as tasklets
+- **Current Location**: Embedded in target-specific copy generation
+
+#### 12. **SDFGSplittingPass**
+- **Purpose**: Split complex SDFGs into multiple files if needed
+- **Input**: SDFG with memlet lowering complete
+- **Output**: List of SDFGs (one per target file)
+- **Current Location**: Implicit in current system
+- **Note**: Moved earlier to enable per-file frame and target code generation
+
+#### 13. **TargetCodeGenerationPass**
+- **Purpose**: Generate both frame code and target-specific code for each SDFG file by calling the appropriate target code generators
+- **Input**: Split SDFGs with all previous analysis
+- **Output**: pipeline_results["code_objects"] = List[CodeObject] with complete code
+- **Current Location**: Combined from `DaCeCodeGenerator.generate_code()` and target-specific `get_generated_codeobjects()`
+- **Note**: This pass calls the target code generators (CppCodeGen, GPUCodeGen, FPGACodeGen, etc.) to generate platform-specific code
+
+#### 14. **HeaderGenerationPass**
+- **Purpose**: Generate C/C++ header files for SDFG interface
+- **Input**: CodeObjects with complete code
+- **Output**: pipeline_results["headers"] = {call_header, sample_main}
+- **Current Location**: `generate_headers()`, `generate_dummy()`
+
+## Information Flow Design
+
+### Main Code Generation Pipeline
+
+```python
+class CodeGenerationPipeline(Pipeline):
+    """Complete code generation pipeline for DaCe SDFGs."""
+
+    def __init__(self):
+        super().__init__([
+            # Phase 1: Analysis
+            TypeInferencePass(),
+            LibraryExpansionPass(),
+            MetadataCollectionPass(),
+            AllocationAnalysisPass(),
+            ControlFlowAnalysisPass(),
+            TargetAnalysisPass(),
+
+            # Phase 2: Transformations
+            AllocationPass(),
+            CopyToMapPass(),
+            ConditionalPipeline([
+                (lambda r: 'cuda' in r.get('targets', []), StreamAssignmentPass()),
+                (lambda r: 'fpga' in r.get('targets', []), FPGAPreprocessingPass()),
```

Comment:
The FPGAPreprocessingPass was not mentioned above. Are we also going to have an OpenMP-specific pass?